torch_xla/core/xla_builder.py

```diff
@@ -855,32 +860,141 @@ def jax_func_to_xla_computation(jax_func, args, kwargs, name=None):
   # If we don't do this before calling jax, any torch_xla operation will hang.
   jax_import_guard()
 
-  import jax
-  import torchax.ops.mappings as mappings
-
-  flattened, spec = tree_flatten((args, kwargs))
-
-  def fn_flattened_inputs(*flattened):
-    args, kwargs = tree_unflatten(flattened, spec)
-    return jax_func(*args, **kwargs)
-
-  sample_input_shapes = tuple(
-      jax.ShapeDtypeStruct(a.shape, mappings.t2j_dtype(a.dtype))
-      for a in flattened)
-  # `as_serialized_hlo_module_proto` is mentioned at
-  # https://github.com/jax-ml/jax/discussions/22266
-  hlo_module = jax.jit(
-      fn_flattened_inputs,
-      keep_unused=True).lower(*sample_input_shapes).compiler_ir(
-          'hlo').as_serialized_hlo_module_proto()  # type: ignore
-
-  return XlaComputation(name, hlo_module, flattened)
-
-
-def call_jax(jax_func, args, kwargs=None, name=None):
+  # Prevent JAX from discovering MegaScale devices a second time. If we don't do this,
+  # then the MegaScale device discovery will hang.
+  with _jax_env_context():
+    import jax
+    import torchax.ops.mappings as mappings
+
+    flattened_inputs, spec = jax.tree.flatten((args, kwargs))
+
+    def abstractify(a):  # make a pytree leaf abstract
+      import jax
+      import torch_xla
+      if a is None:
+        return None
+      if isinstance(a, torch.Tensor):
+        assert a.device == torch_xla.device(
+        ), f"Inputs must be XLA tensors. Got {a.device}"
+        return jax.ShapeDtypeStruct(a.shape, mappings.t2j_dtype(a.dtype))
+      return a
+
+    sample_inputs = tuple(abstractify(a) for a in flattened_inputs)
+
+    # Pick out the non-static args.
+    # Consider anything that is not a `jax.ShapeDtypeStruct` as a static arg.
+    def pick_tensor_args(flattened_args):
+      tensor_args = []
+      for i in range(len(sample_inputs)):
+        if isinstance(sample_inputs[i], jax.ShapeDtypeStruct):
+          tensor_args.append(flattened_args[i])
+      return tensor_args
+
+    sample_tensor_args = pick_tensor_args(sample_inputs)
+    tensor_args = pick_tensor_args(flattened_inputs)
+
+    # This function only takes in tensor arguments because its signature must
+    # match the signature of the HLO module lowered from JAX, allowing us to
+    # wrap it in an XLA user computation.
+    def fn(*tensor_args):
+      # Go from a list of tensor args to the full list of flattened arguments,
+      # by referencing the original flattened inputs.
+      new_flattened = copy(flattened_inputs)
```

Comment:
Looks like what you want to do is

`pytree.tree_map_only(torch.Tensor, fn, (args, kwargs))`